// sqreen/lib/enums/addresses.js
/**
 * Copyright (c) 2016 - 2021 Sqreen. All Rights Reserved.
 * Please refer to our terms for more information: https://www.sqreen.io/terms.html
 */
'use strict';
module.exports = {
    GRAPHQL_QUERY: 'server.request.graphql.resolver_calls',
    HTTP_REQUEST_BODY: 'server.request.body',
    HTTP_REQUEST_QUERY: 'server.request.query',
    HTTP_REQUEST_HEADERS: 'server.request.headers.no_cookies',
    HTTP_REQUEST_COOKIES: 'server.request.cookies',
    HTTP_REQUEST_URI: 'server.request.uri.raw',
    HTTP_REQUEST_URI_PARAMS: 'server.request.path_params',
    HTTP_REQUEST_ENDPOINT: 'server.request.framework_endpoint',
    HTTP_REQUEST_BODY_FILES_FIELD_NAMES: 'server.request.body.files_field_names',
    HTTP_REQUEST_BODY_FILES_NAMES: 'server.request.body.filenames',
    HTTP_REQUEST_BODY_FILES_SIZE: 'server.request.body.combined_file_size',
    HTTP_RESPONSE_STATUS: 'server.response.status',
    HTTP_RESPONSE_RAW_BODY: 'server.response.body.raw',
    HTTP_RESPONSE_HEADERS: 'server.response.headers.no_cookies'
};
