// sq-ecosystem/test/test_utils/index.js
/**
 * Copyright (c) 2016 - 2020 Sqreen. All Rights Reserved.
 * Please refer to our terms for more information: https://www.sqreen.io/terms.html
 */
'use strict';

const INTERFACES = require('../../lib/enums').INTERFACES;
const Eco = require('../../lib/interfaces');

const NOT_IMPLEMENTED = function () {

    const err = new Error('mock not implemented');
    console.log(err);
    throw err;
};

module.exports.getEcosystemMock = function () {

    const val = {
        getInterface(name) {

            return this[name];
        },
        [INTERFACES.ENGINE]: {
            subscribe: NOT_IMPLEMENTED,
            createSpanFromRoot: NOT_IMPLEMENTED,
            getNeededFieldSet: NOT_IMPLEMENTED,
            resolveAddress: NOT_IMPLEMENTED
        },
        [INTERFACES.TRANSPORT]: {
            enableHTTP: NOT_IMPLEMENTED,
            getLegacySqreenMiddleware: NOT_IMPLEMENTED,
            getBlockingStatus: NOT_IMPLEMENTED,
            sqreenBlocking: NOT_IMPLEMENTED,
            utils: {
                getXFFOrRemoteAddress: NOT_IMPLEMENTED,
                ensureProperIP: NOT_IMPLEMENTED
            },
            ipAndPath: {
                ACTIONS: {
                    PREVENT: 'prevent',
                    NONE: 'none'
                },
                checkIPandPath: NOT_IMPLEMENTED
            },
            getHttpTrace: NOT_IMPLEMENTED,
            unrefHttpTrace: NOT_IMPLEMENTED,
            stopIncomingTransaction: NOT_IMPLEMENTED,
            stopHttpTransaction: NOT_IMPLEMENTED,
            getSession: NOT_IMPLEMENTED,
            startIncomingTransaction: NOT_IMPLEMENTED,
            propagate: NOT_IMPLEMENTED,
            shouldPropagate: NOT_IMPLEMENTED
        },
        [INTERFACES.INSTRUMENTATION]: {
            loader: null,
            strategies: {
                patchEventListeners: NOT_IMPLEMENTED,
                patchFunction: NOT_IMPLEMENTED,
                massWrap: NOT_IMPLEMENTED,
                wrap: NOT_IMPLEMENTED
            },
            setWrapCallback: NOT_IMPLEMENTED,
            registerHttpServerForReveal: NOT_IMPLEMENTED
        },
        [INTERFACES.AGENT]: {
            reportError: NOT_IMPLEMENTED
        },
        [INTERFACES.TRACING]: {
            shouldTrace: NOT_IMPLEMENTED,
            getTracingIdentifier: NOT_IMPLEMENTED,
            getAsyncStorage: NOT_IMPLEMENTED
        },
        [INTERFACES.SIGNAL]: {
            createPoint: NOT_IMPLEMENTED
        }
    };
    Eco.ecosystemInterface = val;
    return val;
};
