// @snyk/fix/dist/plugins/python/handlers/pip-requirements/update-dependencies/generate-pins.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generatePins = void 0;
const calculate_relevant_fixes_1 = require("./calculate-relevant-fixes");
const is_defined_1 = require("./is-defined");
function generatePins(requirements, updates) {
    // Lowercase the upgrades object. This might be overly defensive, given that
    // we control this input internally, but its a low cost guard rail. Outputs a
    // mapping of upgrade to -> from, instead of the nested upgradeTo object.
    const lowerCasedPins = calculate_relevant_fixes_1.calculateRelevantFixes(requirements, updates, 'transitive-pins');
    if (Object.keys(lowerCasedPins).length === 0) {
        return {
            pinnedRequirements: [],
            changes: [],
            appliedRemediation: [],
        };
    }
    const appliedRemediation = [];
    const changes = [];
    const pinnedRequirements = Object.keys(lowerCasedPins)
        .map((pkgNameAtVersion) => {
        const [pkgName, version] = pkgNameAtVersion.split('@');
        const newVersion = lowerCasedPins[pkgNameAtVersion].split('@')[1];
        const newRequirement = `${pkgName}>=${newVersion}`;
        changes.push({
            success: true,
            userMessage: `Pinned ${pkgName} from ${version} to ${newVersion}`,
        });
        appliedRemediation.push(pkgNameAtVersion);
        return `${newRequirement} # not directly required, pinned by Snyk to avoid a vulnerability`;
    })
        .filter(is_defined_1.isDefined);
    return {
        pinnedRequirements,
        changes,
        appliedRemediation,
    };
}
exports.generatePins = generatePins;
