// sqreen-sdk/lib/sdk.js
'use strict';
// Low level
const Signal = module.exports.Signal = require('./signal');
module.exports.Trace = require('./trace');
module.exports.Point = require('./point');
module.exports.Metric = require('./metric');
const Batch = module.exports.Batch = require('./batch');
const BE = require('./backend');
module.exports.reportSignal = BE.reportSignal;
module.exports.reportTrace = BE.reportTrace;
module.exports.reportBatch = BE.reportBatch;

// Medium Level

module.exports.setReport = function (reportFn) {

    Signal.prototype.report = reportFn;
};

const initBatch = module.exports.initBatch = function (maxBatchSize, maxBatchAge, reportFn) {

    if (reportFn) {
        Batch._report = reportFn;
    }
    Signal.prototype.BATCH = new Batch(maxBatchSize, maxBatchAge, reportFn);
};

module.exports.init = function (token, maxBatchSize, maxBatchAge) {

    BE.setAPIKey(token);
    initBatch(maxBatchSize, maxBatchAge);
};

module.exports.HTTPTrace = require('./traces/httpTrace');

// High level API

module.exports.ExpressMiddleware = require('./frameworks').express;
