// sqreen-sdk/lib/trace.js
'use strict';
const Signal = require('./signal');
const Point = require('./point');
const Metric = require('./metric');

module.exports = class Trace extends Signal {

    /**
     *
     * @param {Date?} time
     */
    constructor(time) {

        super(undefined, undefined, undefined, time);
        this.data = [];
        this.location = undefined;
        this.payload_schema = undefined;
        this.payload = undefined;
        this.trigger = undefined;
    }

    /**
     *
     * @param {string} signalName
     * @param {string} source
     * @param {?} payload
     * @param {Date?} time
     * @return {Point}
     */
    addPoint(signalName, source, payload, time) {

        const point = new Point(signalName, source, time);
        point.payload = payload;
        this.data.push(point);
        return point;
    }

    /**
     *
     * @param {string} name
     * @param {string} source
     * @param {number} period
     * @param {Date} startDate
     * @param {Date} endDate
     * @param {?} values
     * @return {Metric}
     */
    addMetric(name, source, period, startDate, endDate, values) {

        const metric = new Metric(name, source, period, startDate, endDate, values);
        this.data.push(metric);
        return metric;
    }
};
