// sq-ecosystem/lib/enums.js
/**
 * Copyright (c) 2016 - 2020 Sqreen. All Rights Reserved.
 * Please refer to our terms for more information: https://www.sqreen.io/terms.html
 */
'use strict';

module.exports.HEADERS = {
    X_SQREEN_TRACKING: 'x-sqreen-tracking'
};

module.exports.INTERFACES = {
    INSTRUMENTATION: 'instrumentation',
    TRANSPORT: 'transport',
    COMMAND: 'command',
    AGENT: 'agent',
    TRACING: 'tracing',
    SIGNAL: 'signal',
    ENGINE: 'engine'
};

module.exports.TRANSPORT = {
    SCOPE: {
        CLIENT: 'client',
        SERVER: 'server',
        PRODUCER: 'producer',
        CONSUMER: 'consumer',
        HTTP: { // Legacy, TODO: remove
            OUTGOING: 'client',
            INCOMING: 'server'
        }
    },
    TRANSPORT_TYPE: {
        HTTP: 'http',
        HTTPS: 'https',
        REDIS: 'redis',
        SQLITE: 'sqlite',
        MONGO: 'mongo',
        MYSQL: 'mysql',
        PG: 'postgres',
        KAFKA: 'kafka',
        SQS: 'aws-sqs',
        KINESIS: 'aws-kinesis',
        AMQP: 'amqp'
    }
};

module.exports.TRACING = {
    SIGNAL: {
        CONSUMER: {
            NAME: 'tracing.consumer',
            PAYLOAD_SCHEMA: 'tracing/consumer-2020-04-21'
        },
        PRODUCER: {
            NAME: 'tracing.producer',
            PAYLOAD_SCHEMA: 'tracing/producer-2020-04-21'
        },
        SERVER: {
            NAME: 'tracing.server',
            PAYLOAD_SCHEMA: 'tracing/server-2020-04-21'
        },
        CLIENT: {
            NAME: 'tracing.client',
            PAYLOAD_SCHEMA: 'tracing/client-2020-04-21'
        }
    }
};

module.exports.NS_NAME = 'sqreen_ecosystem';
