// sq-ecosystem/lib/transports.js
'use strict';
const Net = require('net');

const INTERFACES = require('./enums').INTERFACES;

const Transport = class {


    addMessageType(fields, value) {

        if (fields.indexOf('message_type') > -1) {
            this.message_type = value;
        }
    }
    /**
     *
     * @param  {string[]} fields
     * @param eInterface
     */
    addTA(fields, eInterface) {

        if (fields.indexOf('tracing_identifier') > -1) {
            this.tracing_identifier = eInterface
                .getInterface(INTERFACES.TRACING)
                .getTracingIdentifier();
        }
    }

    /**
     *
     * @param  {string[]} fields
     * @param {string} ip
     */
    addIP(fields, ip) {

        if (fields.indexOf('ip') > -1 && Net.isIP(ip) > 0) {
            this.ip = ip;
        }
    }

    /**
     *
     * @param  {string[]} fields
     * @param {string} value
     */
    addTransport(fields, value) {

        if (fields.indexOf('transport') > -1) {
            this.transport = value;
        }
    }
};
module.exports.ClientTransport = class extends Transport {};
module.exports.ServerTransport = class extends Transport {};
module.exports.ProducerTransport = class extends Transport {};
module.exports.ConsumerTransport = class extends Transport {};
