// snyk/dist/cli/commands/test/iac-local-execution/results-formatter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FailedToFormatResults = exports.formatScanResults = void 0;
const types_1 = require("./types");
const common_1 = require("../../../../lib/snyk-test/common");
const constants_1 = require("../../../../lib/iac/constants");
const custom_error_1 = require("./../../../../lib/errors/custom-error");
// import {
//   issuesToLineNumbers,
//   CloudConfigFileTypes,
// } from '@snyk/cloud-config-parser';
const SEVERITIES = [common_1.SEVERITY.LOW, common_1.SEVERITY.MEDIUM, common_1.SEVERITY.HIGH];
function formatScanResults(scanResults, options) {
    try {
        // Relevant only for multi-doc yaml files
        const scannedResultsGroupedByDocId = groupMultiDocResults(scanResults);
        const formattedResults = scannedResultsGroupedByDocId.map((iacScanResult) => formatScanResult(iacScanResult, options.severityThreshold));
        return formattedResults;
    }
    catch (e) {
        throw new FailedToFormatResults();
    }
}
exports.formatScanResults = formatScanResults;
//
// function getFileTypeForLineNumber(
//   fileType: string,
// ): CloudConfigFileTypes {
//   switch (fileType) {
//     case 'yaml':
//     case 'yml':
//       return CloudConfigFileTypes.YAML;
//     case 'json':
//       return CloudConfigFileTypes.JSON;
//     default:
//       return CloudConfigFileTypes.YAML;
//   }
// }
const engineTypeToProjectType = {
    [types_1.EngineType.Kubernetes]: constants_1.IacProjectType.K8S,
    [types_1.EngineType.Terraform]: constants_1.IacProjectType.TERRAFORM,
};
function formatScanResult(scanResult, severityThreshold) {
    const formattedIssues = scanResult.violatedPolicies.map((policy) => {
        // TODO: make sure we handle this issue with annotations:
        // https://github.com/snyk/registry/pull/17277
        const cloudConfigPath = scanResult.docId !== undefined
            ? [`[DocId:${scanResult.docId}]`].concat(policy.msg.split('.'))
            : policy.msg.split('.');
        const lineNumber = -1;
        // TODO: once package becomes public, restore the commented out code for having the issue-to-line-number functionality
        // try {
        //   lineNumber = issuesToLineNumbers(
        //     iacFileScanResult.fileContent,
        //     getFileTypeForLineNumber(iacFileScanResult.fileType),
        //     cloudConfigPath,
        //   );
        // } catch (err) {
        //   //
        // }
        return Object.assign(Object.assign({}, policy), { id: policy.publicId, name: policy.title, cloudConfigPath, isIgnored: false, iacDescription: {
                issue: policy.issue,
                impact: policy.impact,
                resolve: policy.resolve,
            }, severity: policy.severity, lineNumber: lineNumber });
    });
    return {
        result: {
            cloudConfigResults: filterPoliciesBySeverity(formattedIssues, severityThreshold),
        },
        isPrivate: true,
        packageManager: engineTypeToProjectType[scanResult.engineType],
        targetFile: scanResult.filePath,
    };
}
function groupMultiDocResults(scanResults) {
    const groupedData = scanResults.reduce((memo, result) => {
        if (memo[result.filePath]) {
            memo[result.filePath].violatedPolicies = memo[result.filePath].violatedPolicies.concat(result.violatedPolicies);
        }
        else {
            memo[result.filePath] = result;
        }
        return memo;
    }, {});
    return Object.values(groupedData);
}
function filterPoliciesBySeverity(violatedPolicies, severityThreshold) {
    if (!severityThreshold || severityThreshold === common_1.SEVERITY.LOW) {
        return violatedPolicies;
    }
    const severitiesToInclude = SEVERITIES.slice(SEVERITIES.indexOf(severityThreshold));
    return violatedPolicies.filter((policy) => severitiesToInclude.includes(policy.severity));
}
class FailedToFormatResults extends custom_error_1.CustomError {
    constructor(message) {
        super(message || 'Failed to format results');
        this.code = types_1.IaCErrorCodes.FailedToFormatResults;
        this.userMessage =
            'We failed printing the results, please contact support@snyk.io';
    }
}
exports.FailedToFormatResults = FailedToFormatResults;
