// sq-ecosystem/manifest.js
/**
 * Copyright (c) 2016 - 2020 Sqreen. All Rights Reserved.
 * Please refer to our terms for more information: https://www.sqreen.io/terms.html
 */
'use strict';
module.exports = {
    transport: {
        enabled: true,
        producer: {
            'kafka-node': true,
            'amqplib': true,
            'aws-sdk': true // FIXME: this does not scale
        },
        consumer: {
            'aws-kcl': true,
            amqplib: true,
            'aws-sdk': true,
            'kafka-node': true
        },
        incoming: {
            http: true,
            https: true,
            express: true,
            fastify: true,
            koa: false,
            'koa-compose': false,
            'koa-body': false,
            'koa-route': false,
            graphql: true,
            'express-graphql': false, // TODO: remove - replaced by "graphql"
            'apollo-server-core': false // TODO: remove - replaced by "graphql"
        },
        outgoing: {
            http: true,
            https: true,
            redis: true,
            sqlite3: true,
            mysql: true,
            mysql2: true,
            pg: true,
            mongodb: true
        }
    }
};
