// @snyk/fix/dist/plugins/python/handlers/pip-requirements/is-supported.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.partitionByFixable = exports.isSupported = exports.projectTypeSupported = void 0;
const chalk = require("chalk");
const contains_require_directive_1 = require("./contains-require-directive");
function projectTypeSupported(res) {
    return !('reason' in res);
}
exports.projectTypeSupported = projectTypeSupported;
async function isSupported(entity) {
    const remediationData = entity.testResult.remediation;
    if (!remediationData) {
        return { supported: false, reason: 'No remediation data available' };
    }
    if (!remediationData.pin || Object.keys(remediationData.pin).length === 0) {
        return {
            supported: false,
            reason: 'There is no actionable remediation to apply',
        };
    }
    // TODO: fix the non null assertion here
    let requirementsTxt;
    try {
        const fileName = entity.scanResult.identity.targetFile;
        requirementsTxt = await entity.workspace.readFile(fileName);
    }
    catch (e) {
        return {
            supported: false,
            reason: e.message,
        };
    }
    const { containsRequire, matches } = await contains_require_directive_1.containsRequireDirective(requirementsTxt);
    if (containsRequire && matches.some((m) => m.includes('c'))) {
        return {
            supported: false,
            reason: `Requirements with ${chalk.bold('-c')} directive are not yet supported`,
        };
    }
    return { supported: true };
}
exports.isSupported = isSupported;
async function partitionByFixable(entities) {
    const fixable = [];
    const skipped = [];
    for (const entity of entities) {
        const isSupportedResponse = await isSupported(entity);
        if (projectTypeSupported(isSupportedResponse)) {
            fixable.push(entity);
            continue;
        }
        skipped.push({
            original: entity,
            userMessage: isSupportedResponse.reason,
        });
    }
    return { fixable, skipped };
}
exports.partitionByFixable = partitionByFixable;
