// sq-ecosystem/test/lib/tracing/client.js
/**
 * Copyright (c) 2016 - 2020 Sqreen. All Rights Reserved.
 * Please refer to our terms for more information: https://www.sqreen.io/terms.html
 */
'use strict';
const Code = require('code');
const Lab = require('lab');
const lab = exports.lab = Lab.script();

const describe = lab.describe;
const it = lab.it;
const expect = Code.expect;

const HTTPOutgoingTracing = require('../../../lib/tracing/client');

describe('Tracing', () => {

    describe('http.outgoing', () => {

        it('should trace an outgoing http request', { plan: 1 }, (done) => {

            const res = HTTPOutgoingTracing({ transport: 'http', host: 'sqreen.io' });
            expect(res).to.equal({
                signal_name: 'tracing.client',
                payload_schema: 'tracing/client-2020-04-21',
                payload: {
                    transport: 'http',
                    ip: undefined,
                    host: 'sqreen.io',
                    tracing_identifier: undefined
                }
            });
            done();
        });
    });
});
