// sq-ecosystem/test/lib/tracing/index.js
/**
 * Copyright (c) 2016 - 2020 Sqreen. All Rights Reserved.
 * Please refer to our terms for more information: https://www.sqreen.io/terms.html
 */
'use strict';
const Code = require('code');
const Lab = require('lab');
const lab = exports.lab = Lab.script();

const describe = lab.describe;
const it = lab.it;
const expect = Code.expect;

const Tracing = require('../../../lib/tracing');
const HTTPClientTransport = require('../../../lib/transports').ClientTransport;

describe('Tracing', () => {

    it('should broke the tracing instances', { plan: 2 }, (done) => {

        const resolved = new HTTPClientTransport();
        resolved.host =  'sqreen.com';
        const res1 = Tracing.trace(resolved);
        expect(res1).to.equal({
            signal_name: 'tracing.client',
            payload_schema: 'tracing/client-2020-04-21',
            payload: {
                transport: undefined,
                ip: undefined,
                host: 'sqreen.com',
                tracing_identifier: undefined
            }
        });
        const res2 = Tracing.trace('foo', { host: 'sqreen.com' });
        expect(res2).to.equal(null);
        done();
    });
});
