// sq-ecosystem/test/lib/tracing/producer.js
/**
 * Copyright (c) 2016 - 2020 Sqreen. All Rights Reserved.
 * Please refer to our terms for more information: https://www.sqreen.io/terms.html
 */
'use strict';
const Code = require('code');
const Lab = require('lab');
const lab = exports.lab = Lab.script();

const describe = lab.describe;
const it = lab.it;
const expect = Code.expect;

const PrducerTracing = require('../../../lib/tracing/producer');

describe('Tracing', () => {

    describe('Producer', () => {

        it('should trace a producer', { plan: 1 }, (done) => {

            const res = PrducerTracing({ transport: 'kafka', ip: '10.2.6.3', tracing_identifier: 'foo', host: 'hello', topic: 'test' });
            expect(res).to.equal({
                signal_name: 'tracing.producer',
                payload_schema: 'tracing/producer-2020-04-21',
                payload: {
                    transport: 'kafka',
                    host: 'hello',
                    topic: 'test',
                    ip: '10.2.6.3',
                    tracing_identifier: 'foo'
                }
            });
            done();
        });
    });
});
