// sq-ecosystem/test/lib/tracing/server.js
/**
 * Copyright (c) 2016 - 2020 Sqreen. All Rights Reserved.
 * Please refer to our terms for more information: https://www.sqreen.io/terms.html
 */
'use strict';
const Code = require('code');
const Lab = require('lab');
const lab = exports.lab = Lab.script();

const describe = lab.describe;
const it = lab.it;
const expect = Code.expect;

const HTTPIncomingTracing = require('../../../lib/tracing/server');

describe('Tracing', () => {

    describe('http.incoming', () => {

        it('should trace an incoming http request', { plan: 1 }, (done) => {

            const res = HTTPIncomingTracing({
                'server.request.transport': 'http',
                'server.request.client_ip': '10.2.6.3',
                'server.request.tracing_identifier': 'foo'
            });
            expect(res).to.equal({
                signal_name: 'tracing.server',
                payload_schema: 'tracing/server-2020-04-21',
                payload: {
                    transport: 'http',
                    client_ip: '10.2.6.3',
                    tracing_identifier: 'foo'
                }
            });
            done();
        });
    });
});
